src/main/java/org/opensearch/neuralsearch/processor/factory/RerankProcessorFactory.java

```diff
@@ -0,0 +1,153 @@
+ * limitations under the License.
+ */
+package org.opensearch.neuralsearch.processor.factory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import lombok.AllArgsConstructor;
+
+import org.opensearch.neuralsearch.ml.MLCommonsClientAccessor;
+import org.opensearch.neuralsearch.processor.rerank.ContextSourceFetcher;
+import org.opensearch.neuralsearch.processor.rerank.DocumentContextSourceFetcher;
+import org.opensearch.neuralsearch.processor.rerank.QueryContextSourceFetcher;
+import org.opensearch.neuralsearch.processor.rerank.RerankType;
+import org.opensearch.neuralsearch.processor.rerank.TextSimilarityRerankProcessor;
+import org.opensearch.search.pipeline.Processor;
+import org.opensearch.search.pipeline.SearchResponseProcessor;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Factory for rerank processors. Must:
+ * - Instantiate the right kind of rerank processor
+ * - Instantiate the appropriate context source fetchers
+ */
+@AllArgsConstructor
+public class RerankProcessorFactory implements Processor.Factory<SearchResponseProcessor> {
+
+    public static final String RERANK_PROCESSOR_TYPE = "rerank";
+    public static final String CONTEXT_CONFIG_FIELD = "context";
+
+    private final MLCommonsClientAccessor clientAccessor;
+
+    @Override
+    public SearchResponseProcessor create(
+        final Map<String, Processor.Factory<SearchResponseProcessor>> processorFactories,
+        final String tag,
+        final String description,
+        final boolean ignoreFailure,
+        final Map<String, Object> config,
+        final Processor.PipelineContext pipelineContext
+    ) {
+        RerankType type = findRerankType(config);
+        boolean includeQueryContextFetcher = ContextFetcherFactory.shouldIncludeQueryContextFetcher(type);
+        List<ContextSourceFetcher> contextFetchers = ContextFetcherFactory.createFetchers(config, includeQueryContextFetcher);
+        switch (type) {
+            case TEXT_SIMILARITY:
+                @SuppressWarnings("unchecked")
+                Map<String, String> rerankerConfig = (Map<String, String>) config.remove(type.getLabel());
+                String modelId = rerankerConfig.get(TextSimilarityRerankProcessor.MODEL_ID_FIELD);
+                if (modelId == null) {
+                    throw new IllegalArgumentException(
+                        String.format(Locale.ROOT, "%s must be specified", TextSimilarityRerankProcessor.MODEL_ID_FIELD)
+                    );
+                }
+                return new TextSimilarityRerankProcessor(description, tag, ignoreFailure, modelId, contextFetchers, clientAccessor);
+            default:
+                throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, "could not find constructor for reranker type %s", type.getLabel())
+                );
```

Comment:
`"Cannot build reranker type %s"`? This one's a little tricky because we pretty much only hit this path if there was an implementation error - the message is more meant for the programmer adding a rerank type to remind them to add it to this switch statement.

Reply:
The new message sounds good.